server/compare/utils.go

```diff
@@ -0,0 +1,155 @@
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package compare
+
+import (
+	"fmt"
+
+	"github.com/dolthub/go-mysql-server/sql"
+	"github.com/dolthub/go-mysql-server/sql/expression"
+
+	"github.com/dolthub/doltgresql/server/functions/framework"
+	pgtypes "github.com/dolthub/doltgresql/server/types"
+)
+
+// CompareRecords compares two record values using the specified comparison operator, |op| and returns a result
+// indicating if the comparison was true, false, or indeterminate (nil).
+func CompareRecords(ctx *sql.Context, op framework.Operator, v1 interface{}, v2 interface{}) (result any, err error) {
+	leftRecord, rightRecord, err := checkRecordArgs(v1, v2)
+	if err != nil {
+		return nil, err
+	}
+
+	hasNull := false
+	hasEqualFields := true
+
+	for i := 0; i < len(leftRecord); i++ {
+		typ1, ok1 := leftRecord[i].Type.(*pgtypes.DoltgresType)
+		typ2, ok2 := rightRecord[i].Type.(*pgtypes.DoltgresType)
+		if !ok1 {
+			return false, fmt.Errorf("expected a DoltgresType, but got %T", leftRecord[i].Type)
+		} else if !ok2 {
+			return false, fmt.Errorf("expected a DoltgresType, but got %T", rightRecord[i].Type)
+		}
+
+		// NULL values are by definition not comparable, so they need special handling depending
+		// on what type of comparison we're performing.
+		if leftRecord[i].Value == nil || rightRecord[i].Value == nil {
+			switch op {
+			case framework.Operator_BinaryEqual:
+				// If we're comparing for equality, then any presence of a NULL value means
+				// we don't have enough information to determine equality, so return nil.
+				return nil, nil
+
+			case framework.Operator_BinaryLessThan, framework.Operator_BinaryGreaterThan,
+				framework.Operator_BinaryLessOrEqual, framework.Operator_BinaryGreaterOrEqual:
+				// If we haven't seen a prior field with non-equivalent values, then we
+				// don't have enough certainty to make a comparison, so return nil.
+				if hasEqualFields {
+					return nil, nil
+				}
+			}
+
+			// Otherwise, mark that we've seen a NULL and skip over it
+			hasNull = true
+			continue
+		}
+
+		leftLiteral := expression.NewLiteral(leftRecord[i].Value, typ1)
+		rightLiteral := expression.NewLiteral(rightRecord[i].Value, typ2)
+
+		// For >= and <=, we need to distinguish between < and = (and > and =). Records
+		// are compared by evaluating each field, in order of significance, so for >= if
+		// the field is greater than, then we can stop comparing and return true immediately.
+		// If the field is equal, then we need to look at the next field. For this reason,
+		// we have to break >= and <= into separate comparisons for > or < and =.
+		switch op {
+		case framework.Operator_BinaryLessThan, framework.Operator_BinaryLessOrEqual:
```

Comment:
I wonder how this check is actually performed in Postgres. Like does it first use `>=` and then, if that returns true, it then follows up with an `=` check to determine whether to end early or not. One way to do this would be to put notices and override the built-in functions with custom ones, but that's not something that we need to do for this PR. Just perhaps make a note or `TODO` to look a bit more into it, in case it comes up in the future for some reason and that extra hint may be a jumping off point for investigation.